RIOT-OS/RIOT
# netdev_ieeee802154: introduce ops and expose hardware address filter

### Contribution description
EDIT: I decided to change a bit the direction of the PR. I'm introducing 2 features at one now.

### IEEE802.15.4 ops
Following the direction of the lower layer rework #12688 and having technology specific HAL, PHY and MAC layers, this PR introduces a `netdev_ieee802154_ops_t` structure that holds typical operations of IEEE802.15.4 network devices.

These ops are implemented in the drivers and should contain the minimal set of instruction in order to perform the task (e.g a "set_channel" ops shouldn't verify if a channel is valid, only set it in the device).
I'm adding here the first op (`set_hw_addr`), explained below.

In the long term, ieee802154 radios should be operated with the `netdev_ieee802154_t` in order to unify (sub) MAC tasks (CSMA, retransmissions, energy scan, etc).

Regarding the decision of  vtables (function pointers) vs switch-case, the binary is smaller in ROM if the operation is triggered (at least on Cortex-M0+):

Using the NETOPT_AFILTER in get() and set():
```
  text	  data	   bss	   dec	   hex
  24696	   504	  4548	 29748	  7434	
```
Using the ops approach:
```
  text	  data	   bss	   dec	   hex
  24684	   504	  4552	 29740	  742c
```
You can compare the latest commit (https://github.com/RIOT-OS/RIOT/pull/13220/commits/5955d990094596134eae26da7885616e9d8738fa) with the initial implementation (https://github.com/RIOT-OS/RIOT/pull/13220/commits/fb258d1160a4a6c721e1eb8347ebeec315278980)

### Hardware address filtering
Strictly speaking, a radio is not aware of an address (extended, short, panid) unless it supports hardware address filtering. This PR adds some helpers to directly access the hardware address filter of IEEE802.15.4 devices and moves the address set logic to `netdev_ieee802154`.

In current master the process of setting an address goes like this:
1. dev->driver-set with NETOPT_ADDRESS, NETOPT_ADDRESS_LONG or NETOPT_NID
2. Device sets address in hardware accelerator
3. Device calls the SubMAC component (`netdev_ieee802154`) in order to set address variables

With this PR:
1. netdev_ieee802154_set with NETOPT_ADDRESS, NETOPT_ADDRESS_LONG or NETOPT_NID
2. SubMAC (`netdev_ieee802154`) set address variables
3. If the underlying device supports address filtering, set the address there.

I removed all address entries from the IEEE802.15.4 network devices (NETOPT_ADDRESS, NETOPT_ADDRESS_LONG or NETOPT_NID) and implements the `netdev_ieee802154_ops->set_hw_addr` function in devices that support address filtering. For compatibility reasons, calling dev->driver->set with any of the address NETOPTs still work. However, I'm trying to remove the network devices dependency with `netdev_ieee802154`.

Also:
- I removed all dependencies to `netdev_ieee802154` address fields from network devices.
- The CC2420 getter was broken anyway, so this one should fix it
- The sixlowpan bit in short addresses was only being set in network devices, not in the SubMAC (`netdev_ieee802154`). So I suspect some radio were not setting it (e.g `nrf802154`)





### Testing procedure
List of radios:
- [ ] at86rf2xx
- [ ] cc2420
- [ ] mrf24j40
- [ ] cc2538
- [ ] nrf802154
- [ ] kw2xrf

- Check documentation with `make doc`
- Check that all radios can send and receive data right after boot
- Check that all radios can send and receive data after setting short address, long address and PANID




### Issues/PRs references
